packages/chem/quri_parts/chem/utils/excitations.py

```diff
@@ -0,0 +1,164 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#      http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Sequence
+
+from typing_extensions import TypeAlias
+
+from quri_parts.circuit import (
+    LinearMappedUnboundParametricQuantumCircuit,
+    LinearParameterFunction,
+    Parameter,
+    ParameterOrLinearFunction,
+)
+
+#: Represents the set of orbital indices involved in excitation. The first element is
+#: the index of an occupied orbital and the second element is the index of an
+#: unoccupied index.
+SingleExcitation: TypeAlias = tuple[int, int]
+
+#: Represents the set of orbital indices involved in excitation. The first and second
+#: element is the indices of occupied orbitals and the third and fourth element is the
+#: indices of unoccupied orbitals.
+DoubleExcitation: TypeAlias = tuple[int, int, int, int]
+
+
+def excitations(
+    n_spin_orbitals: int, n_fermions: int, delta_sz: float = 0
+) -> tuple[Sequence[SingleExcitation], Sequence[DoubleExcitation]]:
+    """Returns the list of excitation indices for :attr:`n_spin_orbitals` and
+    :attr:`n_fermions` with :attr:`delta_sz` as a tuple of
+    :attr:`SingleExcitation` and :class:`DoubleExcitation`."""
+    sz = [0.5 if (i % 2 == 0) else -0.5 for i in range(n_spin_orbitals)]
+    singles = []
+    doubles = []
+    for r in range(n_fermions):
+        for p in range(n_fermions, n_spin_orbitals):
+            if sz[p] - sz[r] == delta_sz:
+                singles.append((r, p))
+    for s in range(n_fermions - 1):
+        for r in range(s + 1, n_fermions):
+            for q in range(n_fermions, n_spin_orbitals - 1):
+                for p in range(q + 1, n_spin_orbitals):
+                    if sz[p] + sz[q] - sz[r] - sz[s] == delta_sz:
```

Comment:
I'm a bit worried about an error coming from numerical precision when comparing float values.
Should we use `int` for `delta_sz`?

Reply:
But the values are always multiples of 0.5, so there should be no numerical errors.
So I think we can keep it as it is!